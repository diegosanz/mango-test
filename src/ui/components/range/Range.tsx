import React, { FC, useEffect, useState, useRef } from "react";
import InvisibleInput from "../invisible-input/InvisibleInput";
import styled from "styled-components";
import RangeControl from "../range-control/RangeControl";

enum RangeControls {
  MIN,
  MAX,
}

const RangeStyles = styled.div`
  .range {
    display: flex;

    &__boundaries {
      max-width: 100%;
    }

    &__bar {
      flex-grow: 1;
      position: relative;
      display: flex;
      flex-direction: column;
      justify-content: center;

      &__control {
        position: absolute;

        /* TODO: borrar estos colores */
        &.m-max {
          background-color: red;
        }
        &.m-min {
          background-color: blue;
        }
      }

      &__rail {
        border-radius: 5px;
        height: 5px;
        background-color: #000;
      }
    }
  }
`;

interface RangeProps {
  values: number[] | { min: number; max: number };
}

const Range: FC<RangeProps> = ({ values }) => {
  const rangeBarRef = useRef<HTMLDivElement>(null);

  const [rangeState, setRangeState] = useState<{
    inputEditable: boolean;
    limits: { min: number; max: number };
    values: number[];
  }>({
    inputEditable: false,
    limits: { min: 0, max: 0 },
    values: [],
  });

  // const [value, setValue] = useState<{
  //   min: number;
  //   max: number;
  // }>({
  //   min: 0,
  //   max: 0,
  // });

  const [moving, setMoving] = useState<{
    lastActiveControl: RangeControls;
    released: boolean;
  }>({
    lastActiveControl: RangeControls.MIN,
    released: true,
  });

  useEffect(() => {
    if (Array.isArray(values)) {
      const limitsSorted = values.sort((a, b) => a - b);
      setRangeState({
        inputEditable: false,
        limits: {
          min: limitsSorted[0],
          max: limitsSorted[limitsSorted.length - 1],
        },
        values: limitsSorted,
      });
    } else if (!isNaN(values.max) && !isNaN(values.min)) {
      // Validate that in fact `min` is minor than `max`
      const min = Math.min(values.min, values.max);
      const max = Math.max(values.min, values.max);
      let steps: number[] = [];

      for (let i = min; i <= max; i++) {
        steps = [...steps, i];
      }

      setRangeState({
        inputEditable: true,
        limits: {
          min: values.min,
          max: values.max,
        },
        values: steps,
      });
    }
  }, [values]);

  const onStartMoving = (
    ev:
      | React.MouseEvent<HTMLDivElement, globalThis.MouseEvent>
      | React.TouchEvent<HTMLDivElement>,
    el: RangeControls
  ) => {
    setMoving({
      lastActiveControl: el,
      released: false,
    });
  };

  const onMoving = (
    ev:
      | React.MouseEvent<HTMLDivElement, globalThis.MouseEvent>
      | React.TouchEvent<HTMLDivElement>,
    el: RangeControls
  ) => {
    if (moving.lastActiveControl === el && !moving.released) {
      if (rangeBarRef.current) {
        const leftSpace = rangeBarRef.current.offsetLeft;
        const rangeWidth = rangeBarRef.current.clientWidth;
        const clientX: number =
          ev.nativeEvent instanceof TouchEvent
            ? (ev as any).touches[0].clientX
            : ev.nativeEvent instanceof MouseEvent
            ? (ev as any).clientX
            : 0;

        const percentajeReal = 100 / (rangeWidth / (clientX - leftSpace));
        let percentaje = percentajeReal;
        if (percentaje > 100) {
          percentaje = 100;
        } else if (percentaje < 0) {
          percentaje = 0;
        }

        ev.currentTarget.style.left = percentaje + "%";
      }
    }
  };

  const onStopMoving = () => {
    setMoving({
      ...moving,
      released: true,
    });
  };

  return (
    <RangeStyles>
      <div className="range">
        <InvisibleInput
          value={rangeState?.limits.min}
          onChange={(ev) => console.log(ev)}
        />
        <div className="range__bar" ref={rangeBarRef}>
          <RangeControl
            onMouseDown={(ev) => {
              onStartMoving(ev, RangeControls.MIN);
            }}
            onTouchStart={(ev) => {
              onStartMoving(ev, RangeControls.MIN);
            }}
            onMouseMove={(ev) => {
              onMoving(ev, RangeControls.MIN);
            }}
            onTouchMove={(ev) => {
              onMoving(ev, RangeControls.MIN);
            }}
            onMouseOut={onStopMoving}
            onMouseUp={onStopMoving}
            onTouchEnd={onStopMoving}
            onTouchCancel={onStopMoving}
            className="range__bar__control m-min"
            style={
              moving.lastActiveControl === RangeControls.MIN
                ? { zIndex: 1 }
                : { zIndex: 0 }
            }
          />

          <RangeControl
            onMouseDown={(ev) => {
              onStartMoving(ev, RangeControls.MAX);
            }}
            onTouchStart={(ev) => {
              onStartMoving(ev, RangeControls.MAX);
            }}
            onMouseMove={(ev) => {
              onMoving(ev, RangeControls.MAX);
            }}
            onTouchMove={(ev) => {
              onMoving(ev, RangeControls.MAX);
            }}
            onMouseOut={onStopMoving}
            onMouseUp={onStopMoving}
            onTouchEnd={onStopMoving}
            onTouchCancel={onStopMoving}
            className="range__bar__control m-max"
            style={
              moving.lastActiveControl === RangeControls.MAX
                ? { zIndex: 1 }
                : { zIndex: 0 }
            }
          />

          <div className="range__bar__rail"></div>
        </div>
        <InvisibleInput
          value={rangeState?.limits.max}
          onChange={(ev) => console.log(ev)}
        />
      </div>
    </RangeStyles>
  );
};

export default Range;
