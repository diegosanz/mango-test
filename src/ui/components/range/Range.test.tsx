import React from "react";
import { render } from "@testing-library/react";
import TestRenderer from "react-test-renderer";
import Range from "./Range";

test("renders Range", () => {
  render(<Range />);
});

test("snapshot empty range match", () => {
  const tree = TestRenderer.create(<Range />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot MinMax range match", () => {
  const tree = TestRenderer.create(
    <Range value={{ min: 15, max: 89 }} options={{ min: 1, max: 100 }} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot Number[] range match", () => {
  const tree = TestRenderer.create(
    <Range value={{ min: 15, max: 89 }} options={[2, 15, 78, 89]} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot disabled range match", () => {
  const tree = TestRenderer.create(<Range disabled={true} />).toJSON();
  expect(tree).toMatchSnapshot();
});
