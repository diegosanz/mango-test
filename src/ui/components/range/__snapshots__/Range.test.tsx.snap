// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot MinMax range match 1`] = `
<div
  className="sc-dlfnbm bdQNGe"
>
  <div
    className="range null"
  >
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        className="span-input"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
    <div
      className="range__bar"
      onMouseMove={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
    >
      <div
        className="sc-gsTCUz vaZaG range__bar__control m-min"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "14.141414141414142%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="sc-gsTCUz vaZaG range__bar__control m-max"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "88.8888888888889%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="range__bar__rail"
      />
    </div>
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        className="span-input"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
  </div>
</div>
`;

exports[`snapshot Number[] range match 1`] = `
<div
  className="sc-dlfnbm bdQNGe"
>
  <div
    className="range null"
  >
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        className="span-input"
        disabled={true}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
    <div
      className="range__bar"
      onMouseMove={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
    >
      <div
        className="sc-gsTCUz vaZaG range__bar__control m-min"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "14.942528735632184%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="sc-gsTCUz vaZaG range__bar__control m-max"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "100%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="range__bar__rail"
      />
    </div>
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        className="span-input"
        disabled={true}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
  </div>
</div>
`;

exports[`snapshot disabled range match 1`] = `
<div
  className="sc-dlfnbm bdQNGe"
>
  <div
    className="range m-disabled"
  >
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        className="span-input"
        disabled={true}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
    <div
      className="range__bar"
      onMouseMove={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
    >
      <div
        className="sc-gsTCUz vaZaG range__bar__control m-min"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "0%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="sc-gsTCUz vaZaG range__bar__control m-max"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "100%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="range__bar__rail"
      />
    </div>
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        className="span-input"
        disabled={true}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
  </div>
</div>
`;

exports[`snapshot empty range match 1`] = `
<div
  className="sc-dlfnbm bdQNGe"
>
  <div
    className="range null"
  >
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        className="span-input"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
    <div
      className="range__bar"
      onMouseMove={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
    >
      <div
        className="sc-gsTCUz vaZaG range__bar__control m-min"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "0%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="sc-gsTCUz vaZaG range__bar__control m-max"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "100%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="range__bar__rail"
      />
    </div>
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        className="span-input"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
  </div>
</div>
`;
